### PR TITLE
add custom metrics for RBACSync

### PR DESCRIPTION
**Pull request by mllu**
According to https://github.com/cruise-automation/rbacsync/issues/8, trying to add some metrics for `controller` and `gsuite/grouper` first and see how it goes

sample metrics exposed on `/metrics` endpoint
```
# HELP rbacsync_binding_status The number of RoleBindings and ClusterRoleBIndings configured by the controller and their statuses
# TYPE rbacsync_binding_status gauge
rbacsync_binding_status{kind="ClusterRoleBinding",status="BindingConfigured"} 14
rbacsync_binding_status{kind="RoleBinding",status="BindingConfigured"} 28
# HELP rbacsync_config_status The number of RBACSyncConfigs and RBACSyncClusterConfigs and the status of the processed config
# TYPE rbacsync_config_status gauge
rbacsync_config_status{kind="RBACSyncClusterConfig",status="BindingError"} 14
rbacsync_config_status{kind="RBACSyncClusterConfig",status="ConfigEnqueued"} 14
rbacsync_config_status{kind="RBACSyncClusterConfig",status="UnknownGroup"} 14
rbacsync_config_status{kind="RBACSyncConfig",status="ConfigEnqueued"} 14
rbacsync_config_status{kind="RBACSyncConfig",status="UnknownGroup"} 14
# HELP rbacsync_gsuite_client_creation_status Total number of the status of gsuite client creations
# TYPE rbacsync_gsuite_client_creation_status counter
rbacsync_gsuite_client_creation_status 0
# HELP rbacsync_gsuite_members_latency The amount of time the calls to gsuite for group memberships
# TYPE rbacsync_gsuite_members_latency histogram
rbacsync_gsuite_members_latency_bucket{le="0.005"} 0
rbacsync_gsuite_members_latency_bucket{le="0.01"} 0
rbacsync_gsuite_members_latency_bucket{le="0.025"} 0
rbacsync_gsuite_members_latency_bucket{le="0.05"} 0
rbacsync_gsuite_members_latency_bucket{le="0.1"} 0
rbacsync_gsuite_members_latency_bucket{le="0.25"} 0
rbacsync_gsuite_members_latency_bucket{le="0.5"} 0
rbacsync_gsuite_members_latency_bucket{le="1"} 0
rbacsync_gsuite_members_latency_bucket{le="2.5"} 0
rbacsync_gsuite_members_latency_bucket{le="5"} 0
rbacsync_gsuite_members_latency_bucket{le="10"} 0
rbacsync_gsuite_members_latency_bucket{le="+Inf"} 0
rbacsync_gsuite_members_latency_sum 0
rbacsync_gsuite_members_latency_count 0
```

Signed-off-by: Meng-Lin Lu <meng-lin.lu@getcruise.com>

### CODE REVIEW

**Comment by mllu** [comment]
@mattlandis could you help me take a look when you're available, thanks!

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -369,6 +384,7 @@ func (c *Controller) handleConfig(config *rbacsyncv1alpha.RBACSyncConfig) error
 			c.recorder.Eventf(config, corev1.EventTypeWarning,
 				EventReasonBindingError, "%v/%v has no members for group %v",
 				config.Namespace, config.Name, binding.Group)
+			metrics.RBACSyncConfigStatus.WithLabelValues(MetricsRBACSyncConfig, EventReasonBindingError).Inc()
```
Comment (dustin-decker) [inline pkg/controller/controller.go]: Are we certain that this should be an error? Could a group legitimately be empty?

Comment (dustin-decker) [reply pkg/controller/controller.go]: Maybe we need a EventReasonBindingWarning?

File: pkg/groups/gsuite/grouper.go [hunk 2]
```diff
@@ -135,16 +139,24 @@ func (g *Grouper) Members(group string) ([]rbacv1.Subject, error) {
 
 		switch {
 		case isNotFound(err):
+			metrics.RBACSyncGsuiteMembersStatus.WithLabelValues("NotFound").Inc()
 			return nil, errors.Wrapf(groups.ErrNotFound, "gsuite does not have group: %v", err)
 		case isTimeout(tctx):
+			metrics.RBACSyncGsuiteMembersStatus.WithLabelValues("Timeout").Inc()
 			return nil, errors.Wrapf(groups.ErrTimeout, "timeout calling gsuite api: %v", err)
 		case isCanceled(tctx):
+			metrics.RBACSyncGsuiteMembersStatus.WithLabelValues("Cancled").Inc()
```
Comment (dustin-decker) [inline pkg/groups/gsuite/grouper.go]: Cancled -> Canceled

File: pkg/controller/controller.go (hunk 1 above)
Comment (mllu) [reply pkg/controller/controller.go]: sounds good, will update

**Comment by mllu** [comment]
@dustin-decker Thanks for help review, just addressed those comments. Could you help take another look? Thanks in advance!

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -41,23 +40,35 @@ import (
...
+	"github.com/cruise-automation/rbacsync/pkg/metrics"
 )
 
 const (
 	controllerAgentName = "rbacsync"
 )
 
+const (
+	MetricsRBACSyncConfig        = "RBACSyncConfig"
+	MetricsRBACSyncClusterConfig = "RBACSyncClusterConfig"
```
Comment (stevvooe) [inline pkg/controller/controller.go]: Could you make this consistent with the other references? ClusterRBACSyncConfig is what I think we use.

File: pkg/metrics/metrics.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+	RBACSyncGsuiteClientCreationStatus = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "rbacsync_gsuite_client_creation_status",
+		Help: "Total number of the status of gsuite client creations",
+	})
+	RBACSyncGsuiteMembersStatus = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "rbacsync_gsuite_members_status",
+		Help: "Total number of the status of calls to gsuite with labels for state",
+	}, []string{"status"})
+	RBACSyncGsuiteMembersLatency = promauto.NewHistogram(prometheus.HistogramOpts{
+		Name: "rbacsync_gsuite_members_latency",
```
Comment (stevvooe) [inline pkg/metrics/metrics.go]: Prometheus best practice here would to have this suffixed with `duration_seconds`.

**Comment by stevvooe** [comment]
@mllu thanks for the contribution! Looks pretty good. I have a few nits on naming. Also, let’s have @mattlandis take a peek.

File: pkg/controller/controller.go (hunk 3 above)
Comment (mllu) [reply pkg/controller/controller.go]: will do!

File: pkg/metrics/metrics.go (hunk 4 above)
Comment (mllu) [reply pkg/metrics/metrics.go]: updated, thanks for reviewing

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -41,23 +40,35 @@ import (
...
 	"github.com/cruise-automation/rbacsync/pkg/groups"
+	"github.com/cruise-automation/rbacsync/pkg/metrics"
 )
 
 const (
 	controllerAgentName = "rbacsync"
 )
 
+const (
+	MetricsRBACSyncConfig        = "RBACSyncConfig"
```
Comment (mattlandis) [inline pkg/controller/controller.go]: Do these need to be exported?  Or should they be in the metrics package.
If they are in the metrics package they could be prefixed by what they are `metrics.LabelKindRBACSyncConfig`.

File: pkg/groups/gsuite/grouper.go [hunk 6]
```diff
@@ -110,16 +112,18 @@ func (g *Grouper) Members(group string) ([]rbacv1.Subject, error) {
 	client, err := g.service(ctx)
 	if err != nil {
 		return nil, errors.Wrapf(groups.ErrUnknown,
-		"unable to determine group members, an error occurred creating gsuite client: %v",
-		err)
+			"unable to determine group members, an error occurred creating gsuite client: %v",
+			err)
 	}
+	metrics.RBACSyncGsuiteClientCreationStatus.Inc()
```
Comment (mattlandis) [inline pkg/groups/gsuite/grouper.go]: This is not incremented in the error case which I think is more useful.  We can add labels with success or failure (or status codes).  I think users would want to alert on failures of client creation and would care less about the number of successes.

File: pkg/groups/gsuite/grouper.go [hunk 7]
```diff
@@ -135,16 +139,24 @@ func (g *Grouper) Members(group string) ([]rbacv1.Subject, error) {
...
 		case isCanceled(tctx):
+			metrics.RBACSyncGsuiteMembersStatus.WithLabelValues("Canceled").Inc()
 			return nil, errors.Wrapf(groups.ErrCanceled, "the context canceled the call to gsuite: %v", err)
 		default:
+			metrics.RBACSyncGsuiteMembersStatus.WithLabelValues("Unknown").Inc()
 			return nil, errors.Wrapf(groups.ErrUnknown, "error retrieving group members: %v", err)
 		}
 
 	}
+	metrics.RBACSyncGsuiteMembersLatency.Observe(
```
Comment (mattlandis) [inline pkg/groups/gsuite/grouper.go]: This metric is only getting emitted for success. Would it make sense to add a status similar to GsuiteMembersStatus with the latency for those as well? Or do you think that latency only matters in the success case?

File: pkg/groups/gsuite/grouper.go (hunk 6 above)
Comment (mllu) [reply pkg/groups/gsuite/grouper.go]: sounds good

File: pkg/controller/controller.go (hunk 5 above)
Comment (mllu) [reply pkg/controller/controller.go]: moving into metrics package makes more sense, removing `Metrics` prefix as well, thanks for pointing out.

File: pkg/groups/gsuite/grouper.go (hunk 7 above)
Comment (mllu) [reply pkg/groups/gsuite/grouper.go]: yea, I was somehow under impression error came back really fast and probably not important, but realize there is timeout error, so updated to instrument them all. Thanks for the suggestion.

File: pkg/groups/gsuite/grouper.go (hunk 6 above)
Comment (mllu) [reply pkg/groups/gsuite/grouper.go]: updated, thanks for pointing out.

File: pkg/controller/controller.go (hunk 5 above)
Comment (mllu) [reply pkg/controller/controller.go]: curious what `LabelKind` would be, like `metrics.ClusterRBACSyncConfig` and `metrics.RBACSyncConfig` (non-cluster)

Comment (mattlandis) [reply pkg/controller/controller.go]: What I meant was This is a Label value, for the Kind Label.  Does that make sense?  prefixing the consts with what they are used for.  ClusterRBACSyncConfig just from the name I would think `metric.ClusterRBACSyncConfig` was a metric of some sort for that type.

File: pkg/metrics/metrics.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+var (
+	// Metrics for Controller
+	RBACSyncConfigStatus = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "rbacsync_config_status",
+		Help: "The number of RBACSyncConfigs and RBACSyncClusterConfigs and the status of the processed config",
+	}, []string{"kind", "status"})
+	RBACSyncBindingStatus = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "rbacsync_binding_status",
+		Help: "The number of RoleBindings and ClusterRoleBIndings configured by the controller and their statuses",
```
Comment (mattlandis) [inline pkg/metrics/metrics.go]: (nit): Typo BIndings (capital I) -> Bindings.

Comment (mllu) [reply pkg/metrics/metrics.go]: my mistake, updated

File: pkg/controller/controller.go (hunk 5 above)
Comment (mllu) [reply pkg/controller/controller.go]: gotcha

**Comment by mllu** [comment]
@mattlandis I updated those comments, PTAL, thanks!

File: pkg/controller/controller.go (hunk 5 above)
Comment (mllu) [reply pkg/controller/controller.go]: prefixed all label kind metrics with `LabelKind`

**Comment by mllu** [comment]
@mattlandis I added prefix `LabelKind` to all label kind metrics, PTAL, thanks!

**Pull request merged**